lib/core/src/Cardano/Wallet.hs

```diff
@@ -2139,13 +2140,38 @@ derivePublicKey ctx wid role_ ix = db & \DBLayer{..} -> do
   where
     db = ctx ^. dbLayer @IO @s @k
 
+derivePublicKeyShared
+    :: forall ctx s k n.
+        ( HasDBLayer IO s k ctx
+        , SoftDerivation k
+        , s ~ SharedState n k
+        )
+    => ctx
+    -> WalletId
+    -> Role
+    -> DerivationIndex
+    -> ExceptT ErrDerivePublicKey IO (k 'AddressK XPub)
+derivePublicKeyShared ctx wid role_ ix = db & \DBLayer{..} -> do
+    addrIx <- withExceptT ErrDerivePublicKeyInvalidIndex $ guardSoftIndex ix
+
+    cp <- mapExceptT atomically
+        $ withExceptT ErrDerivePublicKeyNoSuchWallet
+        $ withNoSuchWallet wid
+        $ readCheckpoint wid
+
+    let acctK = accountPublicKey $ getState cp
+    let addrK = deriveAddressPublicKey acctK role_ addrIx
+
+    return addrK
+  where
+    db = ctx ^. dbLayer @IO @s @k
```

Comment:
This code looks _very_ similar to `derivePublicKeyShelley`. Both function could be unified rather easily by providing the account key lookup as a function:

```hs
derivePublicKey
    :: forall ctx s k n.
        ( HasDBLayer IO s k ctx
        , SoftDerivation k
        )
    => ctx
    -> WalletId
    -> (s -> k 'AccountK XPub)
    -> Role
    -> DerivationIndex
    -> ExceptT ErrDerivePublicKey IO (k 'AddressK XPub)
derivePublicKey ctx wid getAccountKey role_ ix = ...
```

Reply:
done